File: Simulation/3190번 - 뱀/3190.py
import sys
from collections import deque

input = sys.stdin.readline

def show_board():
    for i in range(n + 1):
        for j in range(n + 1):
            print(board[i][j], end=' ')
        print()
    print()

def solve(board, apple, moves):
    dd = [[0, 1], [1, 0], [0, -1], [-1, 0]]
    snake = deque()
    sec = 0
    s_x, s_y = 1, 1
    s_d = 0
    board[s_x][s_y] = 1
    snake.append((s_x, s_y))
    
    while True:
        sec += 1
        
        s_x += dd[s_d][0]
        s_y += dd[s_d][1]
        
        if s_x <= 0 or s_x > n or s_y <= 0 or s_y > n or board[s_x][s_y]:
            break
        
        snake.append((s_x, s_y))
        board[s_x][s_y] = 1
        if apple[s_x][s_y]:
            apple[s_x][s_y] = False
        else:
            tmp_x, tmp_y = snake.popleft()
            board[tmp_x][tmp_y] = 0
            
        # show_board()
        
        if moves and sec == moves[0][0]:
            turn = moves[0][1]
            moves.popleft()
            s_d = (s_d + turn + 4) % 4
    
    print(sec)

if __name__ == '__main__':
    n = int(input())
    board = [[0] * (n + 1) for _ in range(n + 1)]
    k = int(input())
    apple = [[False] * (n + 1) for _ in range(n + 1)]
    for _ in range(k):
        a, b = map(int, input().split())
        apple[a][b] = True
    l = int(input())
    moves = deque()
    for _ in range(l):
        a, b = input().split()
        moves.append((int(a), 1 if b == 'D' else -1))

    solve(board, apple, moves)

File: Simulation/3190번 - 뱀/3190.cpp
#include <iostream>
#include <vector>
#include <queue>

using namespace std;

#define X first
#define Y second

int n, k, l;
int board[101][101] = { 0, };
int apple[101][101] = { 0, };

int dx[4] = { 0, 1, 0, -1 };
int dy[4] = { 1, 0, -1, 0 };
queue<pair<int, char> > moves;

void input() {
    cin >> n >> k;
    for(int i = 0; i < k; i++) {
        int a, b;
        cin >> a >> b;
        apple[a][b] = 1;
    }
    cin >> l;
    for(int i = 0; i < l; i++) {
        int a;
        char c;
        cin >> a >> c;
        moves.push({a, c});
    }
}

void show() {
    for(int i = 1; i <= n; i++) {
        for(int j = 1; j <= n; j++) {
            cout << board[i][j] << ' ';
        }
        cout << '\n';
    }
    cout << "\n\n";
}

void solve() {
    // 뱀의 위치 및 방향
    int x = 1, y = 1;
    int dir = 0;
    int sec = 0;
    
    // 뱀의 몸과 맵에서의 뱀
    board[1][1] = 1;
    queue<pair<int, int> > snake;
    snake.push({1, 1});
    
    while(1) {
        x += dx[dir];
        y += dy[dir];
        sec++;

        // 맵 밖으로 나가거나 몸을 부딪힘
        if(x <= 0 || x > n || y <= 0 || y > n || board[x][y]) break;
        
        // 머리 옮기기
        snake.push({x, y});
        board[x][y] = 1;
        
        // 사과가 있는지 확인
        if(apple[x][y]) {
            apple[x][y] = 0;
        }
        else {
            pair<int, int> tail = snake.front();
            board[tail.X][tail.Y] = 0;
            snake.pop();
        }
        
        // 맵에서의 뱀 이동하는 것 보기
        // show();
        
        int move_cnt = moves.front().X;
        int turn = (moves.front().Y == 'D' ? 1 : -1);
        if(sec == move_cnt) {
            dir = ((dir + turn + 4) % 4);
            moves.pop();
        }
    }
    
    cout << sec << '\n';
}

int main() {
    ios::sync_with_stdio(0);
    cin.tie(0);
    
    input();
    solve();
    
    return 0;
}